x/cosmwasmpool/gov_test.go

```diff
@@ -0,0 +1,308 @@
+				wasmKeeperParams.CodeUploadAccess.Addresses = append(addressesAllowedCodeUpload, cwPoolModuleAddress.String())
+				s.App.WasmKeeper.SetParams(s.Ctx, wasmKeeperParams)
+			}
+
+			cosmwasmPoolKeeper := s.App.CosmwasmPoolKeeper
+
+			// System under test.
+			codeId, err := cosmwasmPoolKeeper.UploadCodeIdAndWhitelist(s.Ctx, tc.byteCode)
+
+			if tc.expectedErr {
+				s.Require().Error(err)
+				return
+			}
+
+			s.Require().NoError(err)
+
+			// Check that the code id is whitelisted.
+			s.Require().True(cosmwasmPoolKeeper.IsWhitelisted(s.Ctx, codeId))
+
+			// Validate that the code id is the expected one.
+			s.Require().Equal(tc.expectedCodeId, codeId)
+
+			// Validate that the event is emitted.
+			s.AssertEventEmitted(s.Ctx, types.TypeEvtUploadedCosmwasmPoolCode, 1)
+		})
+	}
+}
+
+// TestUploadCodeIdAndWhitelist tests that core proposal logic for migrating cosmwasm pools
+// works as expected. In particular, we test that the two migration strategies work as expected.
+// 1. Migrate given pools to a pre-uploaded code id.
+// 2. Migrate given pools to a new byte code.
+//
+// It does not test specific errors returned as it is non-trivial to set up
+// due to dependency on the wasm keeper and contract keeper that do not export
+// errors correctly.
+//
+// Test vectors considered:
+// 1. Migration to a pre-uploaded code id works as expected and whitelist updated.
+// 2. Migration to a new byte code works as expected and whitelist updated.
+// 3. Migration fails because the code id is not whitelisted.
+// 3. Migration fails because one of the given pool ids does not exist
+// 4. Migration fails because more than the limit of allowed pools is attempted to migrate.
+// 5. Migration fails because pool id list is empty
+// 6. For success cases, tests that relevant event is emitted.
+func (s *CWPoolGovSuite) TestMigrateCosmwasmPools() {
+	// Get valid transmuter code.
+	validTransmuterCode := s.GetContractCode(apptesting.TransmuterContractName)
+
+	const (
+		preUploadCodeIdPlaceholder  uint64 = 1000
+		zeroCodeId                  uint64 = 0
+		defaultPoolCountToPreCreate uint64 = 3
+	)
+
+	var (
+		emptyByteCode           []byte = []byte{}
+		emptyMigrateMsg         []byte = []byte{}
+		defaultPoolIdsToMigrate        = []uint64{1, 2, 3}
+	)
+
+	tests := []struct {
+		name                                     string
+		poolCountToPreCreate                     uint64
+		poolIdsToMigrate                         []uint64
+		newCodeId                                uint64
+		byteCode                                 []byte
+		migrateMsg                               []byte
+		expectedCodeId                           uint64
+		shouldWhitelistCWPoolModuleAccountUpload bool
+		poolIdLimitOverwrite                     uint64
+
+		// TODO: REMOVE THIS
+		// Temporarily skipping this test case as
+		// transmuter does not have migrate entrypoint exposed.
+		// Tracked in: https://github.com/osmosis-labs/osmosis/issues/5329
+		shouldTempSkipCase bool
```

Comment:
Note this and all relevant TODOs below are tracked in: https://github.com/osmosis-labs/osmosis/issues/5329

CC: @iboss-ptk I need a migration entry point in transmuter please to be able to test this. Once added, I can just unskip these tests

Reply:
You can find that [here](https://github.com/osmosis-labs/transmuter/releases/tag/share-token-with-share-state-removed)